### ISSUE TITLE
Question: UITableView like image rendering for entire sheet content?

### ISSUE THREAD

**Issue by liltimtim** [bannzai/SpreadsheetView#150]
This is not really an issue just a question. 

Is it possible to render out the entire spreadsheet view like you can with a UITableView? 

By this I mean currently I'm creating and extension of SpreadsheetView and doing this...
```
public extension SpreadsheetView {
    public func generateRenderViewImage() -> UIImage? {
        let sheet = self
        
        UIGraphicsBeginImageContext(self.contentSize)
//        let previousFrame = self.frame
//        let contentOffset = self.contentOffset
        dump(self.frame)
        dump(self.contentSize)
        dump(self.scrollView)
        sheet.frame = CGRect.init(x: 0, y: 0, width: self.contentSize.width, height: self.contentSize.height)
        sheet.layer.render(in: UIGraphicsGetCurrentContext()!)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
}
```
I do get an image however it is clipped off in the exact same spot as my device's screen and not the contentSize.  I have even dumped out the contentSize vs. the frame and the difference:
contentSize = 1428
frame = 1024

The generated image seems to be always taking the 1024 frame content but never renders past it.  The image itself is the same size as the contentSize however it clips it off and produces white space after.  

Am I doing something incorrectly?

**Comment by kishikawakatsumi** [comment]
Probably you need to wait for the view to be rendered after changing the frame. The view has not been rendered with a new size yet just by changing the frame. @liltimtim

**Comment by liltimtim** [comment]
@kishikawakatsumi I assumed based upon reading the documentation from apple that sheet.layer.render(in: context) would render everything. 

Here's an excerpt but maybe I'm misunderstanding it. 
```
Renders the layer and its sublayers into the specified context.
This method renders directly from the layer tree, ignoring any animations added to the render tree. Renders in the coordinate space of the layer.
```

Could you possibly provide an example of what you mean?  Much appreciated! 

I'll also post an example of what's happening.  It seems that its clipping off some of the cell as well even though technically the cell is rendered in full. 
![screen shot 2017-09-28 at 8 22 56 am](https://user-images.githubusercontent.com/4106065/30966276-60245144-a426-11e7-85e7-6f5097213501.png)

**Comment by kishikawakatsumi** [comment]
@liltimtim It seems intended behavior. SpreadsheetView only draws only the visible area for performance. The visible area is a rectangle with frame size from the coordinate of contentOffset. That is, to render all areas, first set the frame to the same size as `contentSize`, then run the run loop to execute the draw/layout. Then call `layer.render(in: context)` method.

If you'd like to have more specific advice, please share your code (project is best) first.

**Comment by liltimtim** [comment]
@kishikawakatsumi thank you.  Yes I assumed it renders like a collection view or a table view in that it uses dequeuing technique to save memory.  

However, I've modified a similar approach for a UIScrollView 

```
import UIKit
import SpreadsheetView
class ScrollViewSnapshotter: NSObject {
    
    func PDFWithScrollView(scrollview: SpreadsheetView, destination: String) -> Data {
        let pageDimensions = scrollview.bounds
        let pageSize = pageDimensions.size
        let totalSize = scrollview.contentSize
        let numberOfPagesThatFitHorizontally = Int(ceil(totalSize.width / pageSize.width))
        let numberOfPagesThatFitVertically = Int(ceil(totalSize.height / pageSize.height))
        let outputData = NSMutableData()
        UIGraphicsBeginPDFContextToData(outputData, pageDimensions, nil)
        UIGraphicsBeginImageContext(pageDimensions)
        let savedContentOffset = scrollview.contentOffset
        let savedContentInset = scrollview.contentInset
        scrollview.contentInset = UIEdgeInsets.zero
        if let context = UIGraphicsGetCurrentContext()
        {
            for indexHorizontal in 0 ..< numberOfPagesThatFitHorizontally
            {
                for indexVertical in 0 ..< numberOfPagesThatFitVertically
                {
                    
                    /**
                     *  Step 6a: Start a new page.
                     *
                     *          This automatically closes the previous page.
                     *          There's a similar method UIGraphicsBeginPDFPageWithInfo,
                     *          which allows you to configure the rectangle of the page and
                     *          other metadata.
                     */
                    
                    UIGraphicsBeginPDFPage()
                    let offsetHorizontal = CGFloat.init(indexHorizontal) * pageSize.width
                    let offsetVertical = CGFloat(indexVertical) * pageSize.height
                    
                    scrollview.contentOffset = CGPoint.init(x: offsetHorizontal, y: offsetVertical)
                    scrollview.contentOffset = CGPoint.init(x: offsetHorizontal, y: offsetVertical)
                    context.translateBy(x: -offsetHorizontal, y: -offsetVertical)
                    scrollview.layer.render(in: context)
                }
            }
        }

        UIGraphicsEndPDFContext()

//        scrollview.contentInset = savedContentInset
//        scrollview.contentOffset = savedContentOffset

        return outputData as! Data
    }
}
```

I cannot provide the project as the code is not open sourced.  The above modification to the scroll view does indeed move the sheet however the PDF renders similarly too the previous, it clips

![screen shot 2017-10-10 at 2 26 32 pm](https://user-images.githubusercontent.com/4106065/31403413-06f453ee-adc7-11e7-8307-a18b59ed8c00.png)



Its as if I haven't given it any view at all. 

I'm confused as to why the same code works for a scroll view but not necessarily a table view.  
Now if I just do a regular capture of the context as stated above, everything works with the exception that it clips the view to the rendered area.  So I'm stumped. I'm not sure what exactly I'm doing incorrectly.

**Comment by kishikawakatsumi** [comment]
@liltimtim As explained earlier, SpreadsheetView does not render and lay out views until needed for resource savings. that is why. In order to obtain the expected behavior, it is necessary to specify a `frame` of the same size as `contentSize` or to run run loop as appropriate.

**Comment by liltimtim** [comment]
@kishikawakatsumi thank you for the explanation but I don't think you understood me at all.  I understand how dequeuing works as I explained earlier as well about tableviews. 

I found a solution that works and you were not correct either with the frame suggestion.  

The correct solution was to use the rendering context and shift that in sections as to maintain good memory usage and not render everything all at once.  

```
context.translateBy(x: -offsetHorizontal, y: -offsetVertical)
scrollview.layer.render(in: context)
```
This portion of the code was incorrect on my part as I misread the documentation.  By changing the offsetHorizontal and offsetVertical to negative values was incorrect.  It should have been positive values. 

the corrected code 

```
context.translateBy(x: offsetHorizontal, y: offsetVertical)
scrollview.layer.render(in: context)
// we shift it back the amount because this new Swift 3 translate function is additive 
context.translateBy(x: -offsetHorizontal, y: -offsetVertical)
```

These changes will render everything correctly and then stitch everything nicely.  It is also memory efficient because of the technique it uses by rendering in sections rather than all at once like I was trying to do previously. 


If you wish @kishikawakatsumi, I can integrate this with a pull request feature.  I find it nifty to be able to export the entire table as a PDF and I'm certain others may also find this feature useful.
